### DISCUSSION TITLE
Support fixed point decimals in C++ UDF

### DISCUSSION THREAD

**Discussion by hkulyc** [duckdb/duckdb#6871] [Q&A]
I find that right now UDF functions do not support fixed point decimal as argument and return. Instead, it will automatically transform a fixed decimal to double in C++. I wonder if there is a plan to support this in the future. If so, I am willing to help.

**Comment by Tishj** [comment]
That sounds like a bug actually 🤔 
Could you provide a reproducable example of the problem?
That sounds like our argument caster favors double over decimal, which shouldnt be the case

**Comment by hkulyc** [comment]
So based on the documentation [here](https://duckdb.org/docs/api/cpp.html#:~:text=SQLTypeId%3A%3AFLOAT%2C%20SQLTypeId%3A%3ADOUBLE%2C%20SQLTypeId%3A%3ADECIMAL%20%E2%86%92%20double), the template for SQL::DECIMAL should be double. If I take this rule, fixed decimals will become double in C++ in principle. Give an example for this is easy:
```
double identity(double a){
    return a;
}
```
```
con.CreateScalarFunction<double, double>("identity", &identity);
con.Query("CREATE TABLE test (id int, num decimal(32,30) DEFAULT NULL, PRIMARY KEY (id))");
con.Query("insert into test values (1,2.5), (2,3.33), (3, 0.12345678901234567890)");
con.Query("select id, identity(num) from test")->Print();
```
You can see the output will be different from the inserted value.
```
id      identity(num)
INTEGER DOUBLE
[ Rows: 3]
1       2.5
2       3.33
3       0.12345678901234568
```

**Reply by Tishj** [reply]
Ah that functionality, I'd seen the code for it but never the docs, sounds like it's intended behaviour for now, but definitely possible to support in the future :)

Though there are certain challenges with this, since our DECIMAL type has its scale and width embedded in the type
So you would likely receive a int[16|32|64] and have to assume the scale+width from what you set as the argument type

Apart from those, the biggest width decimals (>18 if I remember correctly) are backed by hugeint_t

**Reply by hkulyc** [reply]
If I were to change this default behavior (to expose fixed point decimal to the UDF), where should I start with? Thank you.
